extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,617 @@
+  {
+    TaskGroup taskGroup = taskGroups.get(groupId);
+    Map<Integer, Long> startPartitions = new HashMap<>();
+    Map<Integer, Long> endPartitions = new HashMap<>();
+    StringBuilder sb = new StringBuilder();
+
+    // Generate a string from the partitions and offsets handled by this task to be used in the sequenceName and
+    // availability group. This must be deterministic so that all tasks handling these partitions/offsets, 1) have the
+    // same sequenceName and hence generate the same segment IDs, and 2) have the same availability group and are thus
+    // assigned to different workers for replication. We only assign "active" partitions, so that replicas and retries
+    // operate on the same set of partitions, even if new partitions were created in the middle of creating tasks.
+    // Partitions become marked as active once the task group they are assigned to has a task complete successfully.
+    for (int partition : taskGroup.partitions.keySet()) {
+      PartitionOffsets offsets = taskGroup.partitions.get(partition);
+      if (offsets.active) {
+        startPartitions.put(partition, offsets.startingOffset);
+        endPartitions.put(partition, offsets.endingOffset);
+        sb.append(String.format("+%d(%d-%d)", partition, offsets.startingOffset, offsets.endingOffset));
+      }
+    }
+    String partitionOffsetStr = sb.toString().substring(1);
+
+    String availabilityGroup = Joiner.on("_")
+                                     .join(
+                                         "index_kafka",
+                                         dataSource,
+                                         (config.taskIdShowOffset()
+                                          ? partitionOffsetStr
+                                          : Math.abs(partitionOffsetStr.hashCode()))
+                                     );
+
+    Map<String, String> consumerProperties = Maps.newHashMap(spec.getIoConfig().getConsumerProperties());
+    consumerProperties.put("bootstrap.servers", spec.getIoConfig().getKafkaBrokers());
+
+    KafkaIOConfig ioConfig = new KafkaIOConfig(
+        availabilityGroup,
+        new KafkaPartitions(spec.getIoConfig().getTopic(), startPartitions),
+        new KafkaPartitions(spec.getIoConfig().getTopic(), endPartitions),
+        consumerProperties,
+        true
+    );
+
+    for (int i = 0; i < replicas; i++) {
+      String taskId = Joiner.on("_").join(availabilityGroup, getRandomId());
+      KafkaIndexTask indexTask = new KafkaIndexTask(
+          taskId,
+          new TaskResource(availabilityGroup, 1),
+          spec.getDataSchema(),
+          spec.getTuningConfig(),
+          ioConfig,
+          // used to track which tasks were created by this instance of the supervisor so we can terminate unowned tasks
+          ImmutableMap.<String, Object>of(SUPERVISOR_CONTEXT_KEY, supervisorRunId)
+      );
+
+      if (taskMaster.getTaskQueue().isPresent()) {
```

Comment:
It's better to grab the Optional, check it, then `get()` it. Otherwise you're vulnerable to a race where you lose leadership between the call to `isPresent()` and the call to `get()`.